compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs

```diff
@@ -386,20 +499,89 @@ simplifyExpr world = fst . cata go
       -- NOTE(MH): This also works when `x` is free in `e2` since let-bindings
       -- are _not_ recursive.
       ETmAppF (ETmLam (x, t) e1, s0) (e2, s2) ->
-        go $ ELetF (BindingF (x, t) (e2, s2)) (e1, s1)
+        go world $ ELetF (BindingF (x, t) (e2, s2)) (e1, s1)
         where
           s1 = infoUnstepETmapp x s0
 
       -- (let x = e1 in e2) e3    ==>    let x = e1 in e2 e3, if x is not free in e3
       ETmAppF (ELet (Binding (x, t) e1) e2, s0) e3
         | not (isFreeExprVar x (freeVars (snd e3))) ->
-          go $ ELetF (BindingF (x, t) (e1, s1)) (go $ ETmAppF (e2, s2) e3)
+          go world $ ELetF (BindingF (x, t) (e1, s1)) (go world $ ETmAppF (e2, s2) e3)
           where
             (s1, s2) = infoUnstepELet x s0
 
       -- e    ==>    e
       e -> (embed (fmap fst e), infoStep world (fmap snd e))
 
+-- | If we have a closed term under a lambda, we want to lift it up to the top level,
+-- even though the result of the lambda is also a closed term. Similarly, if there
+-- are branches of a case pattern that are closed, we choose to lift them to the top
+-- level.
+alwaysLiftUnder :: ExprF t -> Bool
+alwaysLiftUnder = \case
+    ETyLamF _ _ -> True
+    ETmLamF _ _ -> True
+    ECaseF _ _ -> True
+    _ -> False
+
+-- | Some terms are not worth lifting to the top level, because they don't
+-- require any computation.
+isWorthLifting :: Expr -> Bool
+isWorthLifting = \case
+    EVar _ -> False
+    EVal _ -> False
+    EBuiltin _ -> False
+    EEnumCon _ _ -> False
+    ENil _ -> False
+    ENone _ -> False
+    ELocation _ e -> isWorthLifting e
+    _ -> True
+
+data SimplifierState = SimplifierState
+    { sWorld :: World
+    , sVersion :: Version
+    , sModule :: Module
+    , sReserved :: Set.Set ExprValName
+    , sCache :: Map.Map Expr ExprValName
```

Comment:
Maybe a hashmap would be a better idea given that an `Expr` can be quite large?

Reply:
Maybe. I'll look into it, but I'll leave it as a separate PR.